compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -0,0 +1,2197 @@
+        }
+
+        updateReadOnlyAndNullableFlag(unionType);
+        symResolver.markParameterizedType(unionType, memberTypes);
+        resolvingTypes.pop();
+        return unionType;
+    }
+
+    private void updateReadOnlyAndNullableFlag(BUnionType type) {
+        LinkedHashSet<BType> memberTypes = type.getMemberTypes();
+        LinkedHashSet<BType> flattenMemberTypes = new LinkedHashSet<>(memberTypes.size());
+        boolean isImmutable = true;
+        boolean hasNilableType = false;
+
+        for (BType memBType : BUnionType.toFlatTypeSet(memberTypes)) {
+            if (Types.getReferredType(memBType).tag != TypeTags.NEVER) {
+                flattenMemberTypes.add(memBType);
+            }
+
+            if (isImmutable && !Symbols.isFlagOn(memBType.flags, Flags.READONLY)) {
+                isImmutable = false;
+            }
+        }
+
+        if (isImmutable) {
+            type.flags |= Flags.READONLY;
+            if (type.tsymbol != null) {
+                type.tsymbol.flags |= Flags.READONLY;
+            }
+        }
+
+        for (BType memberType : flattenMemberTypes) {
+            if (memberType.isNullable() && memberType.tag != TypeTags.NIL) {
+                hasNilableType = true;
+                break;
+            }
+        }
+
+        if (hasNilableType) {
+            LinkedHashSet<BType> bTypes = new LinkedHashSet<>(flattenMemberTypes.size());
+            for (BType t : flattenMemberTypes) {
+                if (t.tag != TypeTags.NIL) {
+                    bTypes.add(t);
+                }
+            }
+            flattenMemberTypes = bTypes;
+        }
+
+        for (BType memberType : flattenMemberTypes) {
+            if (memberType.isNullable()) {
+                type.setNullable(true);
+            }
+        }
+        type.setOriginalMemberTypes(memberTypes);
+        memberTypes.clear();
+        memberTypes.addAll(flattenMemberTypes);
+    }
+
+    private BType resolveTypeDesc(BLangIntersectionTypeNode td, AnalyzerData data) {
+        currentDepth = data.depth;
+        List<BLangType> constituentTypeNodes = td.constituentTypeNodes;
+
+        SymbolEnv symEnv = data.env;
+        BTypeSymbol intersectionSymbol = Symbols.createTypeSymbol(SymTag.ARRAY_TYPE, Flags.PUBLIC, Names.EMPTY,
+                symEnv.enclPkg.symbol.pkgID, null, symEnv.scope.owner, td.pos, BUILTIN);
+        BIntersectionType intersectionType = new BIntersectionType(intersectionSymbol);
+        intersectionSymbol.type = intersectionType;
+        LinkedHashSet<BType> constituentTypes = new LinkedHashSet<>();
+        resolvingTypes.push(intersectionType);
+
+        boolean errored = false;
+        int errorTypesCount = 0;
+        int numOfNonReadOnlyConstituents = 0;
+        BLangTypeDefinition typeDefinition = data.typeDefinition;
+        for (BLangType typeNode : constituentTypeNodes) {
+            BType constituentType = resolveTypeDesc(symEnv, typeDefinition, data.depth, typeNode, data);
```

Comment:
For samples like
```bal
type A B & readonly;

type B record {|
    A a;
    int b;
|};
```
We are going through `private BType resolveTypeDesc(BLangIntersectionTypeNode td, AnalyzerData data) {` method twice` This is because when we resolve the constituent type of `A` which is `B`, `A` will be resolved when resolving fields of `B`, hence the fist hit of resolving `A` becomes a duplicate. Can we move this for loop to the top and check whether the intersection type is resolved after resolving the constituent type. If to return the type?

Reply:
Yes. I will update

Reply:
Updated b171cd5faa3088924765ee2b8931e044d195e940